algopy_testing/src/algopy_testing/models/contract.py

```diff
@@ -1,2 +1,78 @@
-class Contract:
-    pass
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, final
+
+if TYPE_CHECKING:
+    import algopy
+
+
+@dataclass
+class StateTotals:
+    global_uints: int | None = None
+    global_bytes: int | None = None
+    local_uints: int | None = None
+    local_bytes: int | None = None
+
+
+class _ContractMeta(type):
+    def __call__(cls, *args: Any, **kwargs: dict[str, Any]) -> object:
+        from algopy import Contract
+
+        from algopy_testing.context import get_test_context
+
+        context = get_test_context()
+        instance = super().__call__(*args, **kwargs)
+
+        if context and isinstance(instance, Contract):
+            context.add_contract(instance)
+
+        return instance
+
+
+class Contract(metaclass=_ContractMeta):
+    """Base class for an Algorand Smart Contract"""
+
+    _name: str
+    _scratch_slots: Any | None
+    _state_totals: StateTotals | None
+
+    def __init_subclass__(
+        cls,
+        *,
+        name: str | None = None,
+        scratch_slots: (
+            algopy.UInt64 | tuple[int | algopy.UInt64, ...] | list[int | algopy.UInt64] | None
+        ) = None,
+        state_totals: StateTotals | None = None,
+    ):
+        cls._name = name or cls.__name__
+        cls._scratch_slots = scratch_slots
+        cls._state_totals = state_totals
+
+    def approval_program(self) -> algopy.UInt64 | bool:
+        # Implement the logic for the approval program
+        from algopy import UInt64
+
+        return UInt64(1)  # Example return value
```

Comment:
These two methods on Contract should be either abstract (but that might be tricky with the custom metaclass) or just raise `NotImplementedError` as the puya compiler does not provide defaults for these methods